#Real-time breast cancer data using ML models

##Purpose of the project

The purpose of the project is to predict 
the breast cancer using 
existing dataset.

##Which ML models are used?

As I researched different Machine Learning models for the breast cancer dataset, it appears that Logistic Regression, Decision Tree Classifier and Random Forest Classifier might suited for this dataset

##which ML model is best suited for given data set


The  breast cancer is a high-dimensions dataset. Among the ML models as i see the Random forest algorithm gives the high accurate value and also used to solve binary classification problems like the breast cancer dataset.


In [ ]:
#loading brest cancer data set from sklearn
import sklearn.datasets
breast_cancer=sklearn.datasets.load_breast_cancer()

In [ ]:
x=breast_cancer.data
y=breast_cancer.target

In [ ]:
#creating a data frame
import pandas as pd
data=pd.DataFrame(breast_cancer.data, columns=breast_cancer.feature_names)

In [ ]:
data['class']=breast_cancer.target

In [ ]:
# we will provide x as input value and y as output value for all these ML models which are given below 
x=data.drop('class',axis=1)
y=data['class']

#Test train split

To train the model we need to provide input and also output values, we have both the values as x and y, After training we need to test the model to find the best out of these models for the given data set, So we have divided our data set using 'Test train split' model


In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.30,random_state=0,stratify=y)

#Logistic Regression

##Experiment 1 -  logistic regresion


"As I researched different machine learning models and checked the breast cancer dataset, it appears that logistic regression might also suited for this dataset. Because, we are providing numerical input data, and logistic regression also takes numerical input. The output we need is in the form of binary (benign or malignant (0 or 1)), and logistic regression provides a binary output as well."


In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
log_model=LogisticRegression()
log_model.fit(x_train,y_train)

LogisticRegression()

In [ ]:
y_pred=log_model.predict(x_test)

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score

In [ ]:
confusion_matrix(y_test,y_pred),accuracy_score(y_test,y_pred)

(array([[ 56,   8],
        [  5, 102]]),
 0.9239766081871345)

#Experiment 2 - Decision tree classifier

Decision tree classifier is a supervised learning algorithm used for predicting a categorical output variable based on one or more input variables.
The advantage of decision trees is that they are easy to interpret and visualize, and they can handle both categorical and numerical data. However, they can suffer from overfitting if the tree is too complex and does not generalize well to new data.

In [ ]:
from sklearn.tree import DecisionTreeClassifier
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.35,random_state=0,stratify=y)

In [ ]:
my_tree=DecisionTreeClassifier()
my_tree.fit(x_train,y_train)
y_pred=my_tree.predict(x_test)

In [ ]:
confusion_matrix(y_test,y_pred),accuracy_score(y_test,y_pred)

(array([[ 68,   7],
        [  8, 117]]),
 0.925)

##Experiment 3 - Random Forest Classifier

Random forest is a versatile machine learning algorithm that can be used for both classification and regression problems.
Random forest is an ensemble learning method we create a random forest classifier model using the RandomForestClassifier class and fit the model to the training data using the fit method. We then use the predict method to make predictions on the test data and calculate the accuracy of the model using the accuracy_score function.

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier()
rf.fit(x_train,y_train)

RandomForestClassifier()

In [ ]:
y_pred=rf.predict(x_test)

In [ ]:
confusion_matrix(y_test,y_pred),accuracy_score(y_test,y_pred)

(array([[ 69,   6],
        [  4, 121]]),
 0.95)

## Conclusion 

After comparing these three machine learning models, we get accuracy as

Logistic Regression        - '0.923'

Desission Tree Calssifier  - '0.925'

Random Forest classifier   - '0.95'

So,we can conclude that the Random Forest Classifier model provides greater accuracy than the other models for the breast cancer dataset.